container_api/src/interface.rs

```diff
@@ -156,10 +189,155 @@ impl ContainerApiBuilder {
             .insert(instance_name.clone(), instance.clone());
         self
     }
+
+    fn unwrap_params_map(params: Params) -> Result<Map<String, Value>, jsonrpc_core::Error> {
+        match params {
+            Params::Map(map) => Ok(map),
+            _ => Err(jsonrpc_core::Error::invalid_params("expected params map")),
+        }
+    }
+
+    fn get_as_string<T: Into<String>>(
+        key: T,
+        params_map: &Map<String, Value>,
+    ) -> Result<String, jsonrpc_core::Error> {
+        let key = key.into();
+        Ok(params_map
+            .get(&key)
+            .ok_or(jsonrpc_core::Error::invalid_params(format!(
+                "`{}` param not provided",
+                &key
+            )))?
+            .as_str()
+            .ok_or(jsonrpc_core::Error::invalid_params(format!(
+                "`{}` is not a valid json string",
+                &key
+            )))?
+            .to_string())
+    }
+
+    pub fn with_admin_dna_functions(mut self) -> Self {
+        self.io
+            .add_method("admin/dna/install_from_file", move |params| {
+                let params_map = Self::unwrap_params_map(params)?;
+
+                let id = Self::get_as_string("id", &params_map)?;
+                let path = Self::get_as_string("path", &params_map)?;
+
+                container_call!(|c| c.install_dna_from_file(PathBuf::from(path), id.to_string()))?;
+
+                Ok(serde_json::Value::String("success".into()))
```

Comment:
From the call this morning it sounds like returning a JSON string messes with the hc-web-client or hc-redux-middleware. We could change this to a full json object like
```
{success: true}
```
but this seems kind of redundant since it is already returning `Ok`. 

I believe problem comes from this line https://github.com/holochain/hc-redux-middleware/blob/e50540d24466c9ef1b4ac9d5dd4c1384b8565197/src/lib/middleware.ts#L25

I will test @thedavidmeister 's suggestion of using a RawString and will see if this can be parsed successfully

Reply:
There is also some inconsistency between 'info/instances' and 'admin/dna/list' as to if they return a stringified result or a regular json result. For an empty config we have
'info/instances': `{"jsonrpc":"2.0","result":"[]","id":"0"}`
'admin/dna/list': `{"jsonrpc":"2.0","result":[],"id":"0"}`

Reply:
As I've pointed out in another comment thread, we want to keep `info/instance` as it is different to `admin/instance/list`. `info/instances/ is available in every interface and listing the instances that are registered in this interface. As long as the interface is not restarted that won't change, yes, but `info/instances` is always correctly showing which instances are there.. ;)